### DISCUSSION TITLE
The cluster has only two nodes. If one node fails, how can the remaining nodes continue to lead the cluster?【moby 23.0.3】

### DISCUSSION THREAD

**Discussion by pigwow** [moby/moby#45871] [Q&A]
The cluster has only two nodes. If one node fails, how can the remaining nodes continue to lead the cluster?【moby 23.0.3】
I know this is unreasonable, but how to modify the moby source code to implement this function? Can someone help me? Thank you.

I made these modifications:
vendor/go.etcd.io/etcd/raft/v3/quorum/majority.go
q := len(c)/2 + 1 --> 
q := len(c)/2 + 1
if len(c) == 2 {
q=1
}

but the error still occurs. The error information is as follows:
level=error msg="Handler for POST /v1.42/nodes/eexh9gekfcb1o2kz7gppsa4e6/update returned error: rpc error: code = Unknown desc = raft: failed to process the request: node lost leader status"
level=error msg="node is no longer leader, aborting propose"

**Comment by thaJeztah** [comment]
Not sure how to patch the code to make that work. Doing so would be fully in conflict with how the [raft algorithm](https://en.wikipedia.org/wiki/Raft_(algorithm)) works. It would mean that 1 manager can modify state without the other managers knowing state changed, and now your cluster would be in an inconsistent state (one manager thinks a service is removed, and another thinks it's not, and happily continues deploying tasks)).

In general; never, ever (EVER!) run an even number of managers; 1 manager is better than 2 (having 2 managers doubles the chance of losing control over your cluster), but 3 is better than one (3 provides redundancy; you can lose 1 manager and still control the cluster).

**Reply by pigwow** [reply]
I know this is unreasonable. However, we have a service scenario where a customer provides only two VMs and needs to ensure that one VM is faulty and the other VM can continue to run properly. Therefore, we want to adapt the source code to a 2-node cluster and select the active node after one node is abnormal. Can you help? Thank you very much.
